File: BiliLiveLib/src/test/java/charlie/bililivelib/LogUtil.java
package charlie.bililivelib;

import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.util.StringTokenizer;

public class LogUtil {
    private static final String LINE_BREAKING = System.lineSeparator();

    private static Logger logger;

    public static void init() {
        logger = LogManager.getLogger(BiliLiveLib.class);
    }

    public static synchronized void logException(Level level, @NotNull @Nls String message, @NotNull Throwable ex) {
        if (logger == null) return;
        logger.log(level, message);
        StringTokenizer tokenizer = getStackTrace(ex);
        while(tokenizer.hasMoreTokens()) {
            logger.log(level, tokenizer.nextToken());
        }
    }

    private static StringTokenizer getStackTrace(Throwable ex) {
        ByteArrayOutputStream temp = new ByteArrayOutputStream();
        ex.printStackTrace(new PrintStream(temp));
        String stackTrace = new String(temp.toByteArray());
        return new StringTokenizer(stackTrace, LINE_BREAKING);
    }

    @Contract(pure = true)
    private static Logger logger() {
        return logger;
    }
}

File: BiliLiveOnlineKeeper/src/main/kotlin/charlie/bilionlinekeeper/util/LogUtil.kt
package charlie.bilionlinekeeper.util

import org.apache.logging.log4j.Level
import org.apache.logging.log4j.Logger
import org.jetbrains.annotations.Contract
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import java.util.*

object LogUtil {
    private val LINE_BREAKING = System.lineSeparator()

    @Synchronized fun logException(logger: Logger, level: Level, message: String, e: Throwable) {
        logger.log(level, message)

        val stackTraceStrings: StringTokenizer = toStackTrace(e)
        while (stackTraceStrings.hasMoreTokens())
            logger.log(level, stackTraceStrings.nextToken())
    }

    @Contract(pure = true)
    private fun toStackTrace(e: Throwable): StringTokenizer {
        val temp = ByteArrayOutputStream()
        e.printStackTrace(PrintStream(temp))
        return StringTokenizer(String(temp.toByteArray()), LINE_BREAKING)
    }
}